### PR TITLE
Jed/makefile optflags

### PR DESCRIPTION
**Pull request by jedbrown**
I think this supersedes #90. @Steven-Roberts Can you confirm that it satisfies your needs?

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -47,21 +46,19 @@ CUDA_DIR  ?= $(or $(patsubst %/,%,$(dir $(patsubst %/,%,$(dir \
...
 
-CFLAGS = -std=c99 -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP
-FFLAGS = -cpp     -Wall -Wextra -Wno-unused-parameter -Wno-unused-dummy-argument -fPIC -MMD -MP
+OPT    = -O -g
+CFLAGS = -std=c99 $(OPT) -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP
+FFLAGS = -cpp     $(OPT) -Wall -Wextra -Wno-unused-parameter -Wno-unused-dummy-argument -fPIC -MMD -MP
+NVCCFLAGS = $(OPT)
 # If using the IBM XL Fortran (xlf) replace FFLAGS appropriately:
 ifneq ($(filter %xlf %xlf_r,$(FC)),)
   FFLAGS = -qpreprocess -qextname -qpic -MMD
```
[the remark this answers is not in the document]
Comment (jedbrown) [reply Makefile]: Thanks, I had missed that that conditional used `=` instead of `+=`. Revised.

**Comment by jeremylt** [comment]
You switched to `-O` instead of `-O2` whereas my PR switched to `-O3` to get loop vectorization. Do we want to add that flag explicitly? I don't know enough about the `O` flag nuts and bolts.

**Comment by jedbrown** [comment]
@jeremylt Chances are you may also want `-march=skylake-avx512 -ffp-contract=fast` and this isn't a business I'd like the makefile to be in. I revised the README to explain this usage.

**Comment by jeremylt** [comment]
Fair enough. So the user will need to find the best flags to get `/cpu/self/opt` to run performative on their specific machine out of the box?

**Comment by jedbrown** [comment]
Yes, but that's typical (especially with different vendor compilers).

**Comment by v-dobrev** [comment]
It is probably best if backends have some control over the build options but that is a discussion better left for another future PR.

**Comment by jedbrown** [comment]
@v-dobrev Perhaps. Providing good defaults across many vendors is messy. Regardless, there should be an easy way for the user to override those defaults.

**Comment by Steven-Roberts** [comment]
@jedbrown yes, this ought to take care of #90.  My only suggestion would be to use += for the original assignment of `CFLAGS`, `FFLAGS`, and `NVCCFLAGS` to allow users a bit more flexibility.

**Comment by jedbrown** [comment]
@Steven-Roberts But if we assign with `+=` then users have no way to override with their own values (and in case of conflicting options, the last one has precedent). I would call that "less flexible".

**Comment by Steven-Roberts** [comment]
I'm not sure I understand.  Wouldn't this allow you to do things like `OPT=-g NVCCFLAGS=-lineinfo make`.  Otherwise, how could you set flags for nvcc only?  I suppose to handle precedence better it should be `CFLAGS := -std=c99 $(OPT) -Wall ... $(CFLAGS)`

**Comment by jedbrown** [comment]
What if `-Wall` is an error for your compiler? The convention with Autotools, CMake, and the like is that if you specify `CFLAGS` then only those flags are used. One could imagine a different convention where specified flags were only supplementary (though a way to say "just use my flags, really" would still be needed), but I don't want to stray too far from convention here.

(Note that `NVCCFLAGS` currently contains a bunch of stuff that should be `NVCPPFLAGS`. It also contains a non-namespaced `-DADD_` which is apparently used by MAGMA, though I don't know why a library would expect something like that in its public interface. Cc: @stomov)

**Comment by Steven-Roberts** [comment]
Ok I'm fine with sticking to convention.  Just to clarify: If I want to set a nvcc specific flag like `-lineinfo`, would that be possible without editing the makefile.

**Comment by jedbrown** [comment]
Yes, but you need to reproduce other necessary options. For example,

    make NVCCFLAGS='--compiler-options=-fPIC -lineinfo'

**Pull request merged**